File: frontend/lib/services/api/list_api_client.dart
import '../../models/group_list.dart';
import 'base_api.dart';

class ListApiClient extends BaseApi {

  Future<List<GroceryList>> fetchLists(String groupId) async {
    return await request<List<GroceryList>>(
      method: 'GET',
      path: '/groups/$groupId/lists',
      fromJson: (json) => (json as List)
          .map((l) => GroceryList.fromJson(l))
          .toList(),
    );
  }

  Future<GroceryList> createList(String name, String groupId) async {
    return await request<GroceryList>(
      method: 'POST',
      path: '/lists',
      body: {
        'name': name,
        'GroupId': groupId,
      },
      fromJson: (json) => GroceryList.fromJson(json),
    );
  }

  Future<void> deleteList(String listId) async {
    return await request<void>(
      method: 'DELETE',
      path: '/lists/$listId',
      fromJson: (_) {},
    );
  }

  Future<GroceryList> archiveList(String listId, String newName) async {
    return await request<GroceryList>(
      method: 'POST',
      path: '/lists/$listId/archive',
      body: {'newName': newName},
      fromJson: (json) => GroceryList.fromJson(json),
    );
  }
}

File: frontend_react/src/services/api/listApiClient.ts
import { BaseApi } from './baseApi';
import {GroceryList} from "@/types/models";


class ListApiClient extends BaseApi {

  /**
   * GET: Fetch all grocery lists associated with a specific group
   */
  async fetchLists(groupId: string): Promise<GroceryList[]> {
    return this.request<GroceryList[]>({
      method: 'GET',
      path: `/groups/${groupId}/lists`,
    });
  }

  /**
   * POST: Create a new grocery list inside a group
   */
  async createList(name: string, groupId: string): Promise<GroceryList> {
    return this.request<GroceryList>({
      method: 'POST',
      path: '/lists',
      body: {
        name,
        GroupId: groupId, // Preserved exact casing from your Dart body ('GroupId')
      },
    });
  }

  /**
   * DELETE: Remove a grocery list by ID
   */
  async deleteList(listId: string): Promise<void> {
    return this.request<void>({
      method: 'DELETE',
      path: `/lists/${listId}`,
    });
  }

  /**
   * POST: Archive an existing list and assign it a new name
   */
  async archiveList(listId: string, newName: string): Promise<GroceryList> {
    return this.request<GroceryList>({
      method: 'POST',
      path: `/lists/${listId}/archive`,
      body: { newName },
    });
  }
}

export const listApiClient = new ListApiClient();